ctapipe/io/serializer.py

```diff
@@ -0,0 +1,356 @@
+
+__all__ = ['FitsSerializer', 'PickleSerializer']
+log.setLevel('INFO')
+
+
+# PICKLE GZIP Implementation
+
+class Serializer(ABC, Component):
+    """
+    Base class for context manager to save Containers to file
+    """
+    def __init__(self, file,  overwrite=False):
+        """
+        Parameters
+        ----------
+        file:  Unicode
+            input or output full path file name
+        overwrite: Bool
+            overwrite outfile file if it already exists
+        """
+        self.file = file
+        self.overwrite = overwrite
+        super().__init__()
+
+    def __enter__(self):
+        log.debug("Serializing on {0}".format(self.file))
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    @abstractmethod
+    def add_container(self, container):
+        """
+        Add a container to serializer
+        """
+        pass
+
+    @abstractmethod
+    def write(self):
+        """
+        Write all containers to outfile
+        """
+        pass
+
+    @abstractmethod
+    def __iter__(self):
+        """
+        Iterate over all containers
+        """
+        pass
+
+
+class PickleSerializer(Serializer):
+    """
+    Serializes list of ctapipe.core.Components.
+    Reads list of Components from file
+    Write list of Components to file
+    """
+    file = Unicode('file name', help='serializer file name').tag(
+        config=True)
+
+    def __init__(self, file=None, overwrite=False):
+        """
+       Parameters
+       ----------
+       file:  Unicode
+          input/output full path file name
+       overwrite: Bool
+           overwrite outfile file if it already exists
+        """
+        if file:
+            self.file = file
+        super().__init__(self.file, overwrite)
+        self.containers = []
+
+    def add_container(self, container):
+        """
+        Add a container to serializer
+        Raises
+        ------
+        TypeError : When container is not type of container
+        """
+        if not isinstance(container, Container):
+            raise TypeError('Can write only Containers')
+        self.containers.append(container)
+
+    def write(self):
+        """
+        Write all containers to outfile
+        Return : str
+             output file with pickle.gzip extension added if it is not already
+              presents.
+        Raises
+        ------
+        PickleError: When containers list cannot be write to outfile
+        PermissionError:
+        FileNotFoundError:
+        """
+        try:
+            file_split = self.file.split('.')
+            if file_split[-1] != 'gzip' or file_split[-2] != 'pickle':
+                output_filename = self.file+'.pickle.gzip'
+            else:
+                output_filename = self.file
+            with gzip.open(output_filename, 'wb') as f_out:
+                dump(self.containers, f_out)
+            return output_filename
+        except (PickleError, PermissionError, FileNotFoundError):
```

Comment:
What's the point of catching the Exception, if you directly reraise it? Did you maybe forget to log the exception?